### ISSUE TITLE
Overlapping styles and entities in convertToHTML not working when intersecting no style

### ISSUE THREAD

**Issue by jebarjonet** [HubSpot/draft-convert#46]
I just noticed the release of #43 which is super cool. Unfortunately, it does not work for intersections situations, which is problematic.

Here is my code: 
```javascript
convertToHTML({
    blockToHTML: { // I put this here to have <div> in my output instead of <p>
        unstyled: {
            start: '<div>',
            end: '</div>',
            empty: '<br>',
        }
    },
    entityToHTML: (entity, originalText) => {
        if (entity.type === 'link') {
            return {
                start: `<a href="${entity.data.url}" target="_blank">`,
                end: '</a>',
            }
        }
        return originalText
    }
})(contentState)
```

### 1. Add link
<img width="65" alt="screen shot 2017-02-22 at 11 49 21" src="https://cloud.githubusercontent.com/assets/4401230/23229486/5b2b228a-f8f5-11e6-85e2-6e5dd884eab2.png">

### 2. Add bold intersecting link
<img width="66" alt="screen shot 2017-02-22 at 11 49 29" src="https://cloud.githubusercontent.com/assets/4401230/23229484/5886ca16-f8f5-11e6-9dd9-376556ccef77.png">

### 3. Result
```html
<div>t<strong>e<</strong>a href="http://google.com" target="_blank">st test</a></div>
```

### 4. Render
<img width="325" alt="screen shot 2017-02-22 at 11 51 33" src="https://cloud.githubusercontent.com/assets/4401230/23229481/54efee32-f8f5-11e6-8512-69a8ace24a28.png">

## Situation

1. Style strictly inside link ✅ 
<img width="85" alt="screen shot 2017-02-22 at 11 58 18" src="https://cloud.githubusercontent.com/assets/4401230/23229766/55f2a9e0-f8f6-11e6-8634-76a39e73094f.png">

2. Style over link ✅ 
<img width="91" alt="screen shot 2017-02-22 at 11 58 11" src="https://cloud.githubusercontent.com/assets/4401230/23229791/6b017582-f8f6-11e6-9d73-c27a83182c08.png">

3. Style outside link ✅ 
<img width="87" alt="screen shot 2017-02-22 at 11 58 21" src="https://cloud.githubusercontent.com/assets/4401230/23229802/7ea56350-f8f6-11e6-856f-9af1edbd3a5e.png">

4. Style intersecting link 🔴 

**Intersect start**

<img width="91" alt="screen shot 2017-02-22 at 12 01 41" src="https://cloud.githubusercontent.com/assets/4401230/23229882/c4e64bf4-f8f6-11e6-8bca-a2d077060c31.png">

Result:

<img width="351" alt="screen shot 2017-02-22 at 12 01 44" src="https://cloud.githubusercontent.com/assets/4401230/23229881/c4e616a2-f8f6-11e6-8b40-9accf212db2d.png">

Code:
```html
<div>t<strong>e<</strong>a href="http://google.com" target="_blank">st test te</a>st</div>
```

**Intersect end**

<img width="81" alt="screen shot 2017-02-22 at 11 56 14" src="https://cloud.githubusercontent.com/assets/4401230/23229821/8ef57218-f8f6-11e6-9bbf-0b53536205ac.png">

Result: 

<img width="112" alt="screen shot 2017-02-22 at 11 58 14" src="https://cloud.githubusercontent.com/assets/4401230/23229822/8ef7a79a-f8f6-11e6-95f0-fbe8632ca547.png">

Code:
```html
<div>te<a href="http://google.com" target="_blank">st test t<strong>e<</strong>/a>st</div>
```

**Comment by kedromelon** [comment]
Should resulting tags overlap, or be close/opened around the intersect? 

In the **Intersect start** example, for example, which of these possibilities would be desired?

- `<div>t<strong>e<a href="http://google.com" target="_blank">s</strong>t test te</a>st</div>
`
- `<div>t<strong>e</strong><a href="http://google.com" target="_blank"><strong>s</strong>t test te</a>st</div>
`

For HTML5, either one should be valid. Is there any preference here?

**Comment by benbriggs** [comment]
I think the latter is what we want here - it's how overlapping style ranges currently work and I'm not so sure the first is valid.

**Comment by benbriggs** [comment]
closed via #57 🎉

**Comment by liyantang** [comment]
@benbriggs Found a breaking case:

Given that you style text not overlapping to links, and have more that one link in text block, style is left on stack and error is thrown. 


Example:

 testing **testing** testing [testing](url) [testing](url)

```
Uncaught Error: Character 73: -1 styles left on stack that should no longer be there
```

Error is thrown in blockInlineStyles: https://github.com/HubSpot/draft-convert/blob/master/src/blockInlineStyles.js#L119

**Comment by benbriggs** [comment]
Hmm I can't seem to reproduce that @liyantang, can you provide a failing test? Here's one I tried to mimic your example that passes:

```
    const contentState = buildContentState([
      {
        type: 'unstyled',
        text: 'testing testing testing testing testing',
        styleRanges: [
          {
            style: 'BOLD',
            offset: 8,
            length: 7
          }
        ],
        entityRanges: [
          {
            key: 0,
            offset: 16,
            length: 7
          },
          {
            key: 1,
            offset: 24,
            length: 7
          }
        ]
      }
    ], {
      0: {
        type: 'LINK',
        mutability: 'MUTABLE',
        data: {
          url: 'asdf'
        }
      },
      1: {
        type: 'LINK',
        mutability: 'MUTABLE',
        data: {
          url: 'asdf'
        }
      }
    });

    const html = convertToHTML({
      entityToHTML: (entity, originalText) => {
        if (entity.type === 'LINK') {
          return <a>{originalText}</a>;
        }
      }
    })(contentState);

    expect(html).toBe('<p>testing <strong>testing</strong> <a>testing</a> <a>testing</a> testing</p>');

**Comment by liyantang** [comment]
```
    it('failing test', () => {
      const contentState = buildContentState([
        {
          type: 'unstyled',
          text: 'overlapping test Hello World',
          styleRanges: [
            {
              offset: 0,
              length: 11,
              style: 'BOLD'
            },
            {
              offset: 23,
              length: 5,
              style: 'ITALIC'
            }
          ],
          entityRanges: [
            {
              key: 0,
              offset: 17,
              length: 5,
              prefixLength: '<a href="http://google.com">'.length,
              suffixLength: '</a>'.length
            },
            {
              key: 1,
              offset: 23,
              length: 5,
              prefixLength: '<a href="http://google.com">'.length,
              suffixLength: '</a>'.length
            }
          ],
        },
      ], {
        0: {
          type: 'LINK',
          mutability: 'IMMUTABLE',
          data: {
            href: 'http://google.com',
          }
        },
        1: {
          type: 'LINK',
          mutability: 'IMMUTABLE',
          data: {
            href: 'http://google.com',
          }
        }
      });

      const result = convertToHTML(convertToHTMLProps)(contentState);

      expect(result).toBe('<strong>overlapping</strong> test <a href="http://google.com">Hello</a> <em><a href="http://google.com">World</a></em>');
    });
```

**Comment by benbriggs** [comment]
Hmm...that still passes for me as well. Is your local fork/clone up to date to `draft-convert@1.4.5`?

**Comment by liyantang** [comment]
I am on the latest of master branch of 1.4.5. I added this test to the existing tests in converToHTML.js and it throws the error for me.

**Comment by benbriggs** [comment]
Strange! Could you try submitting it as a PR so that the test runs in Travis and we can have an isolated place to verify that it can be fixed? Thanks for digging into this!

**Comment by liyantang** [comment]
https://github.com/HubSpot/draft-convert/pull/66

**Comment by benbriggs** [comment]
Perfect, thanks! I'll dig into it now.

**Comment by liyantang** [comment]
Great, thanks!

**Comment by benbriggs** [comment]
Fix is landed in 1.4.6 - thanks so much for your help @liyantang!
